controller/metrics/s3.py

```diff
@@ -0,0 +1,145 @@
+from abc import ABC, abstractmethod
+from dataclasses import dataclass, asdict
+from typing import Optional
+from dateutil import parser
+from datetime import datetime, timedelta
+import boto3
+from smart_open import open as sopen
+import pytz
+from uuid import uuid4
+from pathlib import Path
+import json
+import dataconf
+
+from controller.server_status_enum import ServerStatusEnum
+from controller.metrics.events import MetricEventHandler, MetricEvent
+from controller.metrics.utils import ResourceRequest, resource_request_from_manifest
+
+
+@dataclass
+class S3Config:
+    endpoint: str
+    bucket: str
+    path_prefix: str
+    access_key_id: str
+    secret_access_key: str
+
+    @classmethod
+    def dataconf_from_env(cls, prefix="AUDITLOG_S3_"):
+        return dataconf.env(prefix, cls)
+
+
+@dataclass
+class SesionMetricData:
+    name: str
+    namespace: str
+    uid: str
+    creation_timestamp: datetime
+    resource_request: Optional[ResourceRequest]
+    image: str
+    status: Optional[ServerStatusEnum]
+    old_status: Optional[ServerStatusEnum]
+    commit: Optional[str]
+    repository_url: Optional[str]
+    user: Optional[str]
+
+    def _default_json_serializer(obj):
+        if type(obj) is datetime:
+            return obj.isoformat()
+        if type(obj) is ServerStatusEnum:
+            return obj.value
+
+    def to_json(self):
+        return json.dumps(asdict(self), default=self._default_json_serializer, indent=None)
+
+    @classmethod
+    def from_metric_event(cls, metric_event: MetricEvent):
+        manifest = metric_event.session
+        return cls(
+            manifest.get("metadata", {}).get("name"),
+            manifest.get("metadata", {}).get("namespace"),
+            manifest.get("metadata", {}).get("uid"),
+            parser.isoparse(manifest.get("metadata", {}).get("creationTimestamp")),
+            resource_request_from_manifest(manifest),
+            manifest.get("spec", {}).get("jupyterServer", {}).get("image"),
+            (
+                ServerStatusEnum(manifest.get("status", {}).get("state"))
+                if manifest.get("status", {}).get("state")
+                else None
+            ),
+            metric_event.old_status,
+            manifest.get("metadata", {}).get("annotations", {}).get("renku.io/commit-sha"),
+            manifest.get("metadata", {}).get("annotations", {}).get("renku.io/repository"),
+            manifest.get("metadata", {}).get("annotations", {}).get("renku.io/username"),
+        )
+
+
+class BaseMetricLogger(ABC):
```

Comment:
Did you consider creating a new python `logging` handler for S3 instead? The Info/Debug/Warn model isn't that appropriate for this use case, but it would mean being compatible with many other logging implementations, all the features already available like filtering and I think the basic design of pythons logging is done quite well. At least the shouldRollover/doRollover logic in https://github.com/python/cpython/blob/11e865c74baa45ac9eda69f2f6af05505fc03ce5/Lib/logging/handlers.py#L65 is pretty readable.

I think all that'd be needed is to implement `emit` in a class inheriting from https://github.com/python/cpython/blob/11e865c74baa45ac9eda69f2f6af05505fc03ce5/Lib/logging/__init__.py#L919 and to add the shouldRollover/doRollover logic mentioned above.

But I can see arguments for and against, so just wanted to check if you looked into this. Not something you have to add.

Reply:
Yeah I found the docs on `logging` really hard to navigate. But I am pretty sure I figured things out. I have a logging handler that uploads to s3 on rotation. And I have a formatter as well. But yeah as you suggest using logging for this is much nicer.